lib/srv/app/common/audit.go

```diff
@@ -0,0 +1,243 @@
+		},
+		SessionMetadata: apievents.SessionMetadata{
+			SessionID: identity.RouteToApp.SessionID,
+			WithMFA:   identity.MFAVerified,
+		},
+		UserMetadata: identity.GetUserMetadata(),
+		ConnectionMetadata: apievents.ConnectionMetadata{
+			RemoteAddr: identity.ClientIP,
+		},
+		AppMetadata: apievents.AppMetadata{
+			AppURI:        app.GetURI(),
+			AppPublicAddr: app.GetPublicAddr(),
+			AppName:       app.GetName(),
+		},
+	}
+	return trace.Wrap(a.EmitEvent(ctx, event))
+}
+
+// OnSessionChunk is called when a new session chunk is created.
+func (a *audit) OnSessionChunk(ctx context.Context, sessionCtx *SessionContext, serverID string) error {
+	event := &apievents.AppSessionChunk{
+		Metadata: apievents.Metadata{
+			Type:        events.AppSessionChunkEvent,
+			Code:        events.AppSessionChunkCode,
+			ClusterName: sessionCtx.Identity.RouteToApp.ClusterName,
+		},
+		ServerMetadata: apievents.ServerMetadata{
+			ServerID:        serverID,
+			ServerNamespace: apidefaults.Namespace,
+		},
+		SessionMetadata: apievents.SessionMetadata{
+			SessionID: sessionCtx.Identity.RouteToApp.SessionID,
+			WithMFA:   sessionCtx.Identity.MFAVerified,
+		},
+		UserMetadata: sessionCtx.Identity.GetUserMetadata(),
+		AppMetadata: apievents.AppMetadata{
+			AppURI:        sessionCtx.App.GetURI(),
+			AppPublicAddr: sessionCtx.App.GetPublicAddr(),
+			AppName:       sessionCtx.App.GetName(),
+		},
+		SessionChunkID: sessionCtx.ChunkID,
+	}
+	return trace.Wrap(a.EmitEvent(ctx, event))
+}
+
+// OnRequest is called when an app request is sent during the session and a response is received.
+func (a *audit) OnRequest(ctx context.Context, sessionCtx *SessionContext, req *http.Request, res *http.Response, re *endpoints.ResolvedEndpoint) error {
+	if sessionCtx.App.IsDynamoDB() {
+		return trace.Wrap(a.onDynamoDBRequest(ctx, sessionCtx, req, res, re))
+	}
+	event := &apievents.AppSessionRequest{
+		Metadata: apievents.Metadata{
+			Type: events.AppSessionRequestEvent,
+			Code: events.AppSessionRequestCode,
+		},
+		AppMetadata:        *MakeAppMetadata(sessionCtx.App),
+		Method:             req.Method,
+		Path:               req.URL.Path,
+		RawQuery:           req.URL.RawQuery,
+		StatusCode:         uint32(res.StatusCode),
+		AWSRequestMetadata: *MakeAWSRequestMetadata(req, re),
+	}
+	return trace.Wrap(a.EmitEvent(ctx, event))
+}
+
+// onDynamoDBRequest is called when a DynamoDB app request is sent during the session.
+func (a *audit) onDynamoDBRequest(ctx context.Context, sessionCtx *SessionContext, req *http.Request, res *http.Response, re *endpoints.ResolvedEndpoint) error {
+	// Try to read the body and JSON unmarshal it.
+	// If this fails, we still want to emit the rest of the event info; the request event Body is nullable, so it's ok if body is left nil here.
+	body, err := awsutils.UnmarshalRequestBody(req)
+	if err != nil {
+		a.log.WithError(err).Warn("Failed to read request body as JSON, omitting the body from the audit event.")
```

Comment:
In case of not JSON format can we also attach the req.Body to the `AppSessionDynamoDBRequest` audit event ?

Reply:
the requests to dynamo are all JSON and it needs to be unmarshaled into the protobuf type that expects JSON

Reply:
or are you asking that we have a new field "RawBody" (or some other name) that contains the raw bytes of the request body as a fallback if the body fails to parse as JSON?

Reply:
> the requests to dynamo are all JSON 

All right. I thought that the XML is also supported and in case of XML format we will skip audit event but if we are sure that dynamoDB requests are all JSON that looks fine.

Reply:
See this link: https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Programming.LowLevelAPI.html#Programming.LowLevelAPI.RequestFormat

"The low-level DynamoDB API uses JavaScript Object Notation (JSON) as a wire protocol format."
...
"The AWS SDKs use JSON to send data to DynamoDB, and DynamoDB responds with JSON"